docs/proposals/20230706-yurtappconfigurationreplacement.md

```diff
@@ -0,0 +1,291 @@
+	- [Implementation History](#implementation-history)
+
+## Glossary
+### YurtAppConfigurationReplacement
+YurtAppConfigurationReplacement is a new CRD used to customize the configuration of the workloads managed by YurtAppSet/YurtAppDaemon. It provides an simple and straightforward way to configure every field of the workload under each nodepool. 
+## Summary
+Due to the objective existence of heterogeneous environments such as resource configurations and network topologies in each geographic region, the workload configuration is always different in each region. The workloads(Deployment/StatefulSet) of nodepools in different regions can be rendered through simple configuration by using YurtAppConfigurationReplacement which also supports multiple resources(YurtAppSet/YurtAppDaemon).
+## Motivation
+YurtAppDaemon is proposed for homogeneous workloads. Yurtappset is not user-friendly and scalable, although it can be used for workload configuration by patch field. Therefore, we expect to render different configurations for each workload easily, including replicas, images, configmap, secret, pvc, etc. In addition, it is essential to support rendering of existing resources, like YurtAppSet and YurtAppDaemon, and future resources. 
+### Goals
+- Define the API of YurtAppConfigurationReplacement
+- Provide YurtAppConfigurationReplacement controller
+- Provide Deployment mutating webhook
+- Provide YurtAppConfigurationReplacement validating webhook
+### Non-Goals/Future Work
+- StatefulSet mutating webhook
+- Optimize YurtAppSet(about patch)
+## Proposal
+### Inspiration
+Reference to the design of ClusterRole and ClusterRoleBinding. 
+
+1. Considering the simplicity of customized rendering configuration, an incremental-like approach is used to implement injection, i.e., only the parts that need to be modified need to be declared. They are essentially either some existing resources, such as ConfigMap, Secret, etc., or some custom fields such as Replicas, Env, etc. Therefore, it is reasonable to abstract these configurable fields into an Item. The design of Item refers to the design of VolumeSource in kubernetes. 
+2. In order to inject item into the workloads, we should create a new CRD named YurtAppConfigurationReplacemnet, which consist of replacements. Each replacement replace a set of configuration for matching nodepools. 
+<img src = "../img/yurtappconfigurationreplacement/Inspiration.png" width="600">
+
+### YurtAppConfigurationReplacement API
+1. YurtAppConfigurationReplacement needs to be bound to YurtAppSet/YurtAppDaemon. 
+Considering that there are multiple Deployment/StatefulSet per nodepool, as shown below, it must be bound to YurtAppSet/YurtAppDaemon for injection. We use subject field to bind it to YurtAppSet/YurtAppDaemon. 
+2. YurtAppConfigurationReplacement is responsible for injection of Replacement. We only need to create a new YurtAppConfigurationReplacement resource for all nodepools under a YurtAppSet/YurtAppDaemon resource. 
```

Comment:
Can I think the YurtAppConfigurationReplacement and YurtAppSet is a one-to-one relationship?

Reply:
Yes

Reply:
👌

Reply:
@vie-serendipity I think it is a good idea to add this relationship description in the proposal.